flocker/apiclient/test/test_client.py

```diff
@@ -0,0 +1,181 @@
+                                       maximum_size=DATASET_SIZE,
+                                       metadata={u"hello": u"there"})
+
+        def test_create_conflicting_dataset_id(self):
+            """
+            Creating two datasets with same ``dataset_id`` results in an
+            ``DatasetAlreadyExists``.
+            """
+            client = client_factory()
+            d = self.assert_creates(client, primary=self.node_1,
+                                    maximum_size=DATASET_SIZE)
+
+            def got_result(dataset):
+                d = client.create_dataset(primary=self.node_1,
+                                          maximum_size=DATASET_SIZE,
+                                          dataset_id=dataset.dataset_id)
+                return self.assertFailure(d, DatasetAlreadyExists)
+            d.addCallback(got_result)
+            return d
+
+        def test_move(self):
+            """
+            ``move_dataset`` changes the dataset's primary.
+            """
+            client = client_factory()
+            dataset_id = uuid4()
+
+            d = self.assert_creates(client, primary=self.node_1,
+                                    maximum_size=DATASET_SIZE,
+                                    dataset_id=dataset_id)
+            d.addCallback(
+                lambda _: client.move_dataset(self.node_2, dataset_id))
+
+            def got_result(dataset):
+                listed = client.list_datasets_configuration()
+                listed.addCallback(lambda l: (dataset, l))
+                return listed
+            d.addCallback(got_result)
+
+            d.addCallback(lambda result:
+                          self.assertEqual((result[0], result[0] in result[1]),
+                                           (Dataset(dataset_id=dataset_id,
+                                                    primary=self.node_2,
+                                                    maximum_size=DATASET_SIZE),
+                                            True)))
+            return d
+
+        def test_list_state(self):
+            """
+            ``list_datasets_state`` returns information about state.
+            """
+            client = client_factory()
+            dataset_id = uuid4()
+            d = self.assert_creates(client, primary=self.node_1,
+                                    maximum_size=DATASET_SIZE * 2,
+                                    dataset_id=dataset_id)
+            d.addCallback(lambda _: synchronize_state(client))
+            d.addCallback(lambda _: client.list_datasets_state())
+            d.addCallback(lambda states:
+                          self.assertIn(
+                              DatasetState(dataset_id=dataset_id,
+                                           primary=self.node_1,
+                                           maximum_size=DATASET_SIZE * 2),
+                              states))
+            return d
+
```

Comment:
Those are really tests of the server, it seems? I guess the first one could be something the client checks as well. But probably not essential for proceeding with leases so going to punt on it for now.